packages/node_modules/pouchdb-replication/src/replicate.js

```diff
@@ -131,6 +136,22 @@ function replicate(src, target, opts, returnValue, result) {
       returnValue.emit('change', outResult);
     }
     writingCheckpoint = true;
+
+    var task = src.activeTasks.get(taskId);
```

Comment:
I think there's a possible race condition leading to lost updates in this section of code, because the call to `src.info()` is not synchronised with any of the surrounding code. We shouldn't block on the call to `info()` because this is just done for information purposes and it shouldn't hold up the replication process, but it does mean the task might be updated with a state `completed_items` value.

This would happen if the call to `info()` doesn't resolve quickly, but resolves some time during the next batch, or even later. What we have here is the following:

1. Read the `task` from `activeTasks.get()`
2. An arbitrary amount of time passes
3. Update task via `activeTasks.update()`

It's possible for multiple instances of this process to happen concurrently, meaning the `task` read in step 1 is stale by the time step 3 executes, so multiple instances of step 3 run with the same value of `completed`, causing a lost update.

This can be fixed by making sure you read and update the task in the same event loop tick, i.e. rewrite this section as:

```js
      src.info().then(function (info) {
        if (!currentBatch) {
          return;
        }
        var task = src.activeTasks.get(taskId);
        var completed = task.completed_items || 0;
        var total_items = info.update_seq - initial_last_seq;
        src.activeTasks.update(taskId, {
          completed_items: completed + currentBatch.changes.length,
          total_items
        });
      });
```

The calls to ``activeTasks.get()` and `activeTasks.update()` are now synchronous so the result of the `activeTasks.get()` can't be invalidated by this code executing concurrently.

Reply:
Good point, thanks! I've fixed this with your approach.